lib/memo_wise.rb

```diff
@@ -163,25 +236,46 @@ def self.has_block_arg?(method) # rubocop:disable Naming/PredicateName
   def self.prepended(target) # rubocop:disable Metrics/PerceivedComplexity
     class << target
       # NOTE: See YARD docs for {.memo_wise} directly below this method!
-      def memo_wise(method_name) # rubocop:disable Metrics/PerceivedComplexity
+      def memo_wise(method_name_or_hash) # rubocop:disable Metrics/PerceivedComplexity
+        case method_name_or_hash
+        when Symbol
+          method_name = method_name_or_hash
+        when Hash
+          unless method_name_or_hash.keys == [:self]
+            raise ArgumentError,
+                  "`:self` is the only key allowed in memo_wise"
+          end
+
+          method_name = method_name_or_hash[:self]
+        end
+
+        # In Ruby, "class methods" are implemented as normal instance methods
+        # on the "singleton class" of a given Class object, found via
+        # {Class#singleton_class}.
+        # See: https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91
+        klass = self
+        singleton = klass.singleton_class
+
+        if singleton.private_instance_methods.include?(method_name) ||
```

Comment:
I think all we need to change here as per @ms-ati's observation is something like:

```ruby
if method_name_or_hash.is_a?(Hash) && (
  klass.singleton_class.private_instance_methods.include?(method_name) ||
  klass.singleton_class.instance_methods.include?(method_name)
)
```

and then we can delete all/most of the `case` statement above?

Reply:
If we want to keep throwing an error when a key other than `self:` is passed, I think the `case` statement is actually cleaner. But I tried to make the logic in there more separate and explicit, let me know what you think now!